### DISCUSSION TITLE
8 channel relay with 8 touch sensors

### DISCUSSION THREAD

**Discussion by Arashinof** [HomeSpan/HomeSpan#133] [Q&A] [answered]
Hi im trying to make 3 units, each unit has an esp32 nodemcu and 8 channel relay with 8 push buttons to control on/off manually 
Which example should i go with? I want something simple like example 5 but with on/off button 
I tried to use example 15 but it needs 3 gpios for 3 buttons : 
New DEV_DimmableLED(led pin,1st button,2nd button,3rd button) 
and i dont know what to do, put something like idk "null" for 2nd and 3rd button gpio or maybe some random number like 100 so there is no gpio with this number ?
Or should i go with example 18? 
Is there anything else i should modify too? Cause im planning to make several units for each room and kitchen 

And my second question, im trying to use a ttp223 touch sensor as push button it has 2 modes, active low ( normally low output and High when touched ) And active high ( normally high and Low when touched ) 
How can i modify the code? to disable the internal pull-up because ttp223 normally sends high value in active high mode 
![image](https://user-images.githubusercontent.com/79417601/130151047-1437891f-0d5a-4be8-b477-405dc0d1e104.jpeg)
And this is the wiring i am thinking about ![image](https://user-images.githubusercontent.com/79417601/130162256-e51fd63c-3c7b-44f4-853f-a9e79dbfe814.jpeg)
But instead of switch i want to use ttp223 
Sorry im kinda noob :)
Thanks :)

**Comment by HomeSpan** [comment, marked as answer]
I recommend starting with [Example 5](https://github.com/HomeSpan/HomeSpan/blob/master/docs/Tutorials.md#example-5---workingled) (without any buttons) to get things working.  Then you can layer in either `SpanButton()` logic or `loop()` logic to add in buttons or touch sensors (see my response to your Issue #136).  I create all of my own devices in this piecemeal fashion since it allows me to test and debug each aspect thoroughly before moving to the next step.

**Reply by Arashinof** [reply]
I finally made the circuit and everything works fine except one thing 
My relay module is active low 
Im thinking about connecting all relay outputs to gpios  with internal pullup and Enable pull up on them So the are active high 
And make all the button pins to active low and triggered with High.
By doing this, i have so much more gpios to use 
What should i do? And which part of the code should i make change in? I will be so thankful if you tell me in the amateur way :))

**Comment by HomeSpan** [comment]
No, you cannot enable pull-ups on the GPIOs you use for the relays.  The relays must be connected to pins that are configured as an OUTPUT.  Pull-up resistors are used for INPUTS, not outputs.  In order to flip the logic Example 5 so that HomeSpan outputs a LOW to  "turn on" a relay, and a HIGH to "turns off" a relay, you can simply flip the logic in the line that sets the GPIO pin in the update() method from: 

```C++
digitalWrite(ledPin,power->getNewVal());
```

to

```C++
digitalWrite(ledPin,!power->getNewVal());
```

This is equivalent to:

```C++
if(power->getNewVal()==true)
  digitalWrite(ledPin,LOW)
else
  digitalWrite(ledPin,HIGH)
```

You may also want to change the line:

```C++
power=new Characteristic::On();
```

to 

```C++
power=new Characteristic::On(1);
```

in order to start the device with all relays turn off.

Separately, your touch sensor buttons should each be connected to their own GPIOs (configured as INPUT pins) as per my comments above.

**Reply by Arashinof** [reply]
thanks a lot
i made the changes and its works!! 
but in startup all the relays are on ( outputs are low ) but in the homeapp it shows off 
when i turn the led on and off it goes to active low and works fine 
i think there is something with the startup characteristic; to set all the pins low at startup cause i made the change in 
` power=new Characteristic::On(1); ` but still doesnt works fine at startup till i manually turn the led on and off then its fine 
tried even without the 1 and 0 but still nothing

**Comment by HomeSpan** [comment]
Sorry, my mistake.  Using `power=new Characteristic::On(1)` at startup tells your HomeApp that the Device is ON.  What we needed to do is tell the GPIO pin to startup as HIGH.

To begin, undo the change and revert to `power=new Characteristic::On()`.  Then, add a new line right after the `pinMode(ledPin,OUTPUT);` that says `digitalWrite(ledPin,HIGH);`.  This will cause the pin to be HIGH at start-up, which sets the relay to be off.

I'm not sure if there will be an short flicker where the relay will go on because `pinMode(ledPin,OUTPUT)` automatically starts in LOW mode.  If that happens you can try placing the `digitalWrite` line before the `pinMode` line instead.

**Reply by Arashinof** [reply]
thank you so much 
i tried to modify the bridge example and it works like a charm 
just one more thing 
now im trying to add on/off buttons 
i can use both pull-up and pull-down button with my ttp223 which one do you recommend? 
if i use pulled-up inputs as well as high outputs is it gonna harm my esp? or noise or glitch?
and can i ask you to simplify adding on/off buttons too? or something like an article ( i did read example 12 )  so you dont have to write it for me cause you already helped A LOT!
 thanks :)

**Reply by Arashinof** [reply]
can i use this part of example 18 and try to modify it? 


DEV_DimmableLED(int ledPin, int buttonPin) : Service::LightBulb(){

    power=new Characteristic::On(0,true);           // NEW! Second argument is true, so the value of the On Characteristic (initially set to 0) will be saved in NVS
    level=new Characteristic::Brightness(5,true);   // NEW! Second argument is true, so the value of the Brightness Characteristic (initially set to 5) will be saved in NVS               
    level->setRange(5,100,1);                       // sets the range of the Brightness to be from a min of 5%, to a max of 100%, in steps of 1%

    new SpanButton(buttonPin);                      // create a new SpanButton to control power using PushButton on pin number "buttonPin"

    this->LED=new LedPin(ledPin);                   // configures a PWM LED for output to pin number "ledPin"

    Serial.print("Configuring Dimmable LED: Pin="); // initialization message
    Serial.print(LED->getPin());
    Serial.print("\n");

    LED->set(power->getVal()*level->getVal());      // NEW! IMPORTANT: Set the LED to its initial state at startup.  Note we use getVal() here, since it is set upon instantiation.`
>


i think i need only the `new SpanButton(buttonPin)` part 
right?
and how can i disable homespan config button? can i do it by giving it a gpio that doesnt exis? like gpio 100

**Reply by Arashinof** [reply]
it didnt work

**Comment by HomeSpan** [comment]
My recommendation is to configure your touch buttons to be active LOW, and then implement them exactly as if they were regular pushbuttons as per Example 15.   You can't disable the Control Button - you'll need to assign it to some pin.

**Reply by Arashinof** [reply]
im trying to but my problem is that example 15 and 18 dont have pinmode() in their code

**Comment by HomeSpan** [comment]
No need - if you use SpanButton, all that's built into the HomeSpan code already.  Provided that you can configure your touch buttons to be Active LOW (as would be the case for a pushbutton that connects a pin to ground), HomeSpan should handle the rest.  You only need to del with pinMode, etc. if you want to use a different type of configuration.

**Reply by Arashinof** [reply]
i meant for the output 
and thanks for fast response

**Reply by Arashinof** [reply]
this is my code 
i get the button is pressed message in CLI  but relays dont trigger 





```
#include "extras/PwmPin.h"                          // allows PWM control of LED brightness

struct DEV_LED : Service::LightBulb {               // ON/OFF LED

  int ledPin;                                       // pin number defined for this LED
  int powerPin;
  SpanCharacteristic *power;                        // reference to the On Characteristic
  
  DEV_LED(int ledPin , int powerPin) : Service::LightBulb(){       // constructor() method

    power=new Characteristic::On();                 
    this->ledPin=ledPin; 
    digitalWrite(ledPin,HIGH);                           
    pinMode(ledPin,OUTPUT);                         

    new SpanButton(powerPin);

    this->powerPin=powerPin;                
  } // end constructor

  boolean update(){                              // update() method

    digitalWrite(ledPin,!power->getNewVal());      
   
    return(true);                               // return true
  
  } // update




  void button(int pin, int pressType) override {
   if(pin==powerPin){
       if(pressType==SpanButton::SINGLE){                      // only respond to SINGLE presses
          digitalWrite(ledPin,!power->getNewVal());            // update the physical LED to reflect the new values
          Serial.print("button is pressed");
            }   
   }
  } // butto
}; 
```

**Comment by Arashinof** [comment]
after lots and lots of trying and reading i finally made it 
thanks 
such a great repository 

but i dont know why it freezes and lose connection with accessories when i use CLI commands 
works fine till i input ? in CLI and all accessories in homeapp go "no response"

**Comment by HomeSpan** [comment]
Does the whole system freeze - in other words, if you type 's' into the CLI, nothing is printed?  If so, my guess is that you have inadvertently used one of the serial pins for your touch buttons, or relay, or the control pin.  Eliminate some of your button/relays and see if that solves it - if it starts working again you can determine which button or relay is causing the issue and assign to another pin.

**Reply by Arashinof** [reply]
Im using this config + 2 and 35 for status led and config button 

![image](https://user-images.githubusercontent.com/79417601/130965437-e52f8f29-fa3d-4b6e-bf92-0479bede4ce7.jpeg)

I used 2 serial ports ( gpio 17 and 18 ) as output in my previous config but i changed 
Is there any chance that i broke it? 
Which uart homespan uses? Cause these pins are for uart2 

UPDATE: i flashed the esp32 with a simple program to print serial.read and its fine 
But i cant communicate with it when i flash homespan 

Update2: i erased the flash via esptool and now its ok 
Thanks for your help i really appreciate it

**Comment by gaboskar001** [comment]
@Arashinof could you please share the code?
Thanks in advance

**Reply by niceguy994** [reply]
mine works nearly fine. I just have the problem that sometimes the button is not recognized. So I have to push the button two times an it switches the app & relays. Would be great if we can fix it together.




struct DEV_Switch : Service::Switch {               // First we create a derived class from the HomeSpan LightBulb Service

  
  //LedPin *LED;
  int ledPin;                                       // pin number defined for this LED
  int powerPin;
  SpanCharacteristic *power;                        // reference to the On Characteristic
  
  DEV_Switch (int ledPin , int powerPin) : Service::Switch(){       // constructor() method

    power=new Characteristic::On();                 
    this->ledPin=ledPin;                       
    pinMode(ledPin,OUTPUT); 
    digitalWrite(ledPin,HIGH);                         
    new SpanButton(powerPin);  
    this->powerPin=powerPin;            
  } // end constructor

  boolean update(){                              // update() method

    digitalWrite(ledPin,!power->getNewVal());      
    return(true);                               // return true
  
  } // update



  void button(int powerPin, int pressType) override {
      pinMode(powerPin, INPUT);
       if(pressType==SpanButton::SINGLE){                      // only respond to SINGLE presses
          power->setVal(1-power->getVal());
          digitalWrite(ledPin,!power->getVal());
          LOG1("button is pressed ");
          }           
  }
};

**Reply by Arashinof** [reply]
Can you email me?
I wanted to upload here but apparently cant 
Arashinof@gmail.com

**Comment by gaboskar001** [comment]
Hi @niceguy994 
There is a good chance that contact bounce will occur. Just add a little delay (50ms) after you've read the state of the button. That should solve the problem that you have to press the button sometimes twice.
Can you zip the full code and send it to me? 
I get some errors when  I paste you're code in 'DEV_LED'.
Would be great if we could fix it together.

**Reply by msp1999** [reply]
Hi @Arashinof , @niceguy994  , @gaboskar001  do any of you have a working code for this i was also trying the same thing 
,Just started tinkering with homespan , can you guys please send me the code

**Reply by msp1999** [reply]
issue i am getting is , when i press button it does not updates state in home app
and i had to manually turn it on or off once then it starts working for some time and stops again

**Reply by HomeSpan** [reply]
Are you running with a Home Hub (either an Apple TV or  a HomePod)?

**Reply by msp1999** [reply]
No, i am just using esp32 board as homebridge

**Reply by HomeSpan** [reply]
Starting with about iOS 15 or so, Apple changed HomeKit and an Apple HomePod or TV is required for the Home App to receive notifications from any HomeKit device.  Without one, the Home App closes the connection after a short period of time and HomeSpan can't send any updates.

**Reply by msp1999** [reply]
Ok, thanks for the update

**Comment by liulee027** [comment]
I had a similar problem.
The first press of the button will be unresponsive, and if the button is pressed again within two seconds, the relay can switch states normally, as well as the status of the homekit will be updated.
Responds faster to successive presses.
But after two seconds without pressing the button, the first press after that will be unresponsive.
@HomeSpan

**Reply by liulee027** [reply]
struct HomeSwitch : Service::Switch {               // First we create a derived class from the HomeSpan LightBulb Service

  int SwitchPin;                                       // this variable stores the pin number defined for this LED
  int buttonPin; 
  SpanCharacteristic *power;                        // here we create a generic pointer to a SpanCharacteristic named "power" that we will use below

  // Next we define the constructor for DEV_LED.  Note that it takes one argument, ledPin,
  // which specifies the pin to which the LED is attached.
  HomeSwitch(int SwitchPin,int buttonPin) : Service::Switch(){

    power=new Characteristic::On();                 // this is where we create the On Characterstic we had previously defined in setup().  Save this in the pointer created above, for use below
    this->SwitchPin=SwitchPin;                            // don't forget to store ledPin...
    pinMode(SwitchPin,OUTPUT);                         // ...and set the mode for ledPin to be an OUTPUT (standard Arduino function)
    
    new SpanButton(buttonPin),50000,5,0;
    this->buttonPin=buttonPin;  
  } // end constructor

  // Finally, we over-ride the default update() method with instructions that actually turn on/off the LED.  Note update() returns type boolean

  boolean update(){            
    Serial.println("Hoekit Update");
    digitalWrite(SwitchPin,power->getNewVal());        // use a standard Arduino function to turn on/off ledPin based on the return of a call to power->getNewVal() (see below for more info)
    return(true);                                   // return true to indicate the update was successful (otherwise create code to return false if some reason you could not turn on the LED)
  } // update

  void button(int pin, int pressType) override {
    if(pin==buttonPin){
      if(pressType==SpanButton::SINGLE){   
        Serial.println("Button is Click");       // if a SINGLE press of the power button...
        bool ChangeSwitchToStute = 1-power->getVal();
        Serial.print("Set To =>");
        Serial.println(ChangeSwitchToStute);
        power->setVal(ChangeSwitchToStute);           // ...toggle the value of the power Characteristic
        digitalWrite(SwitchPin,ChangeSwitchToStute);  
      }
    }
  }

};

**Reply by HomeSpan** [reply]
What exactly is not responsive?  Is the relay no longer changing state?

**Reply by liulee027** [reply]
Relay does not respond when first pressed,No messages in the debug window,I'm sure the button is sending a signal.
![Uploading IMG_8823.GIF…]()

**Reply by liulee027** [reply]
It's working now.
`    //new SpanButton(buttonPin);  
    new SpanButton(buttonPin,2000,5,0,SpanButton::TRIGGER_ON_HIGH);`
in the notes “//      SpanButton::TRIGGER_ON_LOW:   used for a button that connects pin to GROUND”
In my schematic，Button IO connected to GND，I don't understand why it's the other way around.

**Comment by CHAITANYAPHADTARE02** [comment]
Hey have you made it ?
